$\Large\textbf{Lab 2. Exercise 2. }$



In [1]:
#numpy package will be used for most of our lab exercises. Please have a look at Please have a look at https://numpy.org/doc/stable/ for numpy documentation
#we will first import the numpy package and name it as np
import numpy as np 
#Henceforth, we can lazily use np to denote the much longer numpy !! 

In [2]:
#Now we will define a function which will compute and return the function value 
def evalf(x):  
  #Input: x is a numpy array of size 2 
  assert type(x) is np.ndarray #do not allow arbitrary arguments 
  #after checking if the argument is valid, we can compute the objective function value
  #compute the function value and return it
  if len(x) == 3:
    return (1/8)*((x[0]-2)**2)+(1/(8**2))*((x[1]-4)**2) + (1/(8**3))*((x[2]-8)**2)
  if len(x) == 4:
    return (1/8)*((x[0]-2)**2)+(1/(8**2))*((x[1]-4)**2) + (1/(8**3))*((x[2]-8)**2) + (1/(8**4))*((x[3]-16)**2)

In [3]:
#Now we will define a function which will compute and return the gradient value as a numpy array 
def evalg(x):  
  #Input: x is a numpy array of size 2 
  assert type(x) is np.ndarray #do not allow arbitrary arguments 
  #after checking if the argument is valid, we can compute the gradient value
  #compute the gradient value and return it 
  #print (len(x))
  #print ("a", np.array([(1/4)*(x[0]-2),(1/32)*(x[1]-4),(1/256)*(x[2]-8),(1/2048)*(x[3]-16)])
  if len(x) == 3:
    return np.array([(1/4)*(x[0]-2),(1/32)*(x[1]-4),(1/256)*(x[2]-8)])
  if len(x) == 4:
    return np.array([(1/4)*(x[0]-2),(1/32)*(x[1]-4),(1/256)*(x[2]-8),(1/2048)*(x[3]-16)])

In [4]:
#Complete the module to compute the steplength by using the closed-form expression
def compute_steplength_exact(gradf, A): #add appropriate arguments to the function 
  assert type(gradf) is np.ndarray #and len(gradf) == 2 
  assert type(A) is np.ndarray #and A.shape[0] == 2 and  A.shape[1] == 2 #allow only a 2x2 array
   
  #Complete the code to compute step length
  #print(gradf)
  #print(A)
  nume=np.matmul(gradf.transpose(),gradf)
  deno=np.matmul(np.matmul(gradf.transpose(),A),gradf)
  step_length=nume/(2*deno)  
  return step_length

In [5]:
#Complete the module to compute the steplength by using the backtracking line search
def compute_steplength_backtracking(x, gradf, alpha_start, rho, gamma): #add appropriate arguments to the function 
  assert type(x) is np.ndarray #and len(gradf) == 2 
  assert type(gradf) is np.ndarray #and len(gradf) == 2 
  
  alpha = alpha_start
  pk=-gradf
  #implement the backtracking line search
  while evalf(x+alpha*pk)>evalf(x)+gamma*alpha*np.matmul(gradf.T,pk):
    alpha=rho*alpha

  #print('final step length:',alpha)
  return alpha

In [6]:
#we define the types of line search methods that we have implemented
EXACT_LINE_SEARCH = 1
BACKTRACKING_LINE_SEARCH = 2
CONSTANT_STEP_LENGTH = 3

In [7]:
def find_minimizer(start_x, tol, line_search_type, *args):
  #Input: start_x is a numpy array of size 2, tol denotes the tolerance and is a positive float value
  assert type(start_x) is np.ndarray #and len(start_x) == 2 #do not allow arbitrary arguments 
  assert type(tol) is float and tol>=0 
  # construct a suitable A matrix for the quadratic function 
  if len(start_x) == 3:
    A = np.array([[1/8, 0, 0],[0, 1/64, 0],[0, 0, 1/512]])
  if len(start_x) == 4:
    A = np.array([[1/8, 0, 0, 0], [0, 1/64, 0, 0], [0, 0, 1/512, 0], [0, 0, 0, 1/4096]])
  x = start_x
  g_x = evalg(x)

  #initialization for backtracking line search
  if(line_search_type == BACKTRACKING_LINE_SEARCH):
    alpha_start = args[0]
    rho = args[1]
    gamma = args[2]
    #print('Params for Backtracking LS: alpha start:', alpha_start, 'rho:', rho,' gamma:', gamma)

  k = 0
  #print('iter:',k, ' x:', x, ' f(x):', evalf(x), ' grad at x:', g_x, ' gradient norm:', np.linalg.norm(g_x))

  while (np.linalg.norm(g_x) > tol): #continue as long as the norm of gradient is not close to zero upto a tolerance tol
  
    if line_search_type == EXACT_LINE_SEARCH:
      #print(g_x)
      step_length = compute_steplength_exact(g_x, A) #call the new function you wrote to compute the steplength
      #raise ValueError('EXACT LINE SEARCH NOT YET IMPLEMENTED')
    elif line_search_type == BACKTRACKING_LINE_SEARCH:
      step_length = compute_steplength_backtracking(x,g_x, alpha_start,rho, gamma) #call the new function you wrote to compute the steplength
      #raise ValueError('BACKTRACKING LINE SEARCH NOT YET IMPLEMENTED')
    elif line_search_type == CONSTANT_STEP_LENGTH: #do a gradient descent with constant step length
      step_length = 0.1
    else:  
      raise ValueError('Line search type unknown. Please check!')
    
    #implement the gradient descent steps here   
    x = np.subtract(x, np.multiply(step_length,g_x)) #update x = x - step_length*g_x
    k += 1 #increment iteration
    g_x = evalg(x) #compute gradient at new point

    #print('iter:',k, ' x:', x, ' f(x):', evalf(x), ' grad at x:', g_x, ' gradient norm:', np.linalg.norm(g_x))
  return x, k, evalf(x)


Code for answer 2

In [8]:
my_start_x=np.array([(1/64),(1/8),1])
my_tol=10**(-10)

In [9]:
x_opt_const, iter_const, f_const = find_minimizer(my_start_x, my_tol, CONSTANT_STEP_LENGTH)
print("Minimizer =", x_opt_const)
print("Minimum function value =", f_const)
print("No of iterations =", iter_const)

Minimizer = [2.         4.         7.99999997]
Minimum function value = 1.2793410906926164e-18
No of iterations = 49723


Code for answer 3

In [10]:
my_start_x=np.array([(1/64),(1/8),1])
my_tol=10**(-10)

In [11]:
x_opt_exact, iter_exact, f_exact = find_minimizer(my_start_x, my_tol, EXACT_LINE_SEARCH)
print("Minimizer =", x_opt_exact)
print("Minimum function value =", f_exact)
print("No of iterations =", iter_exact)

Minimizer = [2.         4.         7.99999998]
Minimum function value = 9.150071377581033e-19
No of iterations = 269


In [12]:
#check what happens when you call find_minimzer using backtracking line search
x_opt_bls, iter_bls, f_bls = find_minimizer(my_start_x, my_tol, BACKTRACKING_LINE_SEARCH, 1, 0.5,0.5)
print("Minimizer =", x_opt_bls)
print("Minimum function value =", f_bls)
print("No of iterations =", iter_bls)

Minimizer = [2.         4.         7.99999997]
Minimum function value = 1.2748574165464873e-18
No of iterations = 4964


Code for answer 4

In [13]:
my_start_x=np.array([(1/512),(1/64),(1/8),1])
my_tol=10**(-10)

In [14]:
x_opt_exact, iter_exact, f_exact = find_minimizer(my_start_x, my_tol, EXACT_LINE_SEARCH)
print("Minimizer =", x_opt_exact)
print("Minimum function value =", f_exact)
print("No of iterations =", iter_exact)

Minimizer = [ 2.          4.          8.         15.99999981]
Minimum function value = 8.8565993523583e-18
No of iterations = 2013


In [15]:
#check what happens when you call find_minimzer using backtracking line search
x_opt_bls, iter_bls, f_bls = find_minimizer(my_start_x, my_tol, BACKTRACKING_LINE_SEARCH, 1, 0.5,0.5)
print("Minimizer =", x_opt_bls)
print("Minimum function value =", f_bls)
print("No of iterations =", iter_bls)

Minimizer = [ 2.         4.         8.        15.9999998]
Minimum function value = 1.0237544252113035e-17
No of iterations = 37079


Answer 2) Minimizer = [2,4,8] and minimum function value = 0

Answer 3) Backtracking line search procedure takes 4964 iterations whereas exact line search procedure takes 269 iterations. 

Answer 4) Minimizer = [2,4,8,16] and minimum function value = 0

Backtracking line search procedure takes 37079 iterations whereas exact line search procedure takes 2013 iterations. 

Answer 5) For any N, Minimizer = [ 2, 4, 8, 16, ... , $2^N$ ] where $i^{th}$ coordinate = $2^i$ and minimum function value = 0.

Usually, backtracking line search takes more iterations than exact line search because of the parameters we specify (alpha, rho and gamma). 

